crates/driver/src/domain/competition/solution/encoding.rs

```diff
@@ -175,12 +183,18 @@ pub fn tx(
                 call_data: interaction.call_data.clone(),
             },
             competition::solution::Interaction::Liquidity(liquidity) => {
-                liquidity_interaction(liquidity, &slippage, contract)?
+                liquidity_interaction(liquidity, &slippage, contracts.settlement())?
             }
         })
     }
 
-    let tx = contract
+    // Encode WETH unwrap
+    if !native_unwrap.0.is_zero() {
+        interactions.push(unwrap(native_unwrap, contracts.weth()));
+    }
```

Comment:
This does not take existing ETH buffers of the settlement contract into account which the current implementation does as a gas optimization. (See https://github.com/gnosis/gp-v2-services/pull/1507 and https://github.com/gnosis/gp-v2-services/issues/1496 as a reference).
I'd be fine with that for the sake of simplicity. Just bringing it up since last estimates suggested that optimization safes around 3% of total gas costs and you might not have been aware.

Reply:
I remember this optimisation (unwrapping all the WETH we have and paying ETH buy orders without unwrapping if we have remaining ETH balance in the contract) but I don't think it's in place right now. I couldn't find any reference to post_processing or optimising in the current codebase. I do think we still merge unwraps but don't unwrap more than we need to or exclude unrwaps from the execution plan.

 E.g. this [settlement]( https://etherscan.io/tx/0xbd409ff0627b6837ceec2a6714a7435aa5f86d71f74d9cb3a89ce9b1dd3c24be) combined 3 unwraps but only unwrapped the exact amount (it could have unwrapped 0.14ETH more) and [this settlement](https://etherscan.io/tx/0x649e7ea9bef666b1d61da19a4d878ce206dda230142e2a658077d72d03d0e885) unwrapped even though the contract had enough balance. So I think this implementation replicates the status quo.

Reply:
I think you are right. I faintly recall disabling this optimization for some reason but can't find the related PR right now. 👌